From 1a655e8951768cec0e508c8bd305765cc64942f1 Mon Sep 17 00:00:00 2001
From: Ananth Krishna R <ananth.krishna.r@intel.com>
Date: Thu, 5 Nov 2015 16:04:54 +0530
Subject: [PATCH] [PCIE]: Ignore PCIE PORT0 child devices in COS state

This patch add neccessary changes to pci/quirks to identify
the platform boot state and also ignore the child devices of
PCIE PORT0 while in Charger OS state. This mechanism can also
be extended to other PCIE root port's based on the need basis.

Change-Id: I46f2885def5049add23a3f0360efb3d420120015
Tracked-On:https://jira01.devtools.intel.com/browse/OAM-7887
Signed-off-by: Ananth Krishna R <ananth.krishna.r@intel.com>
---
 drivers/pci/quirks.c | 29 +++++++++++++++++++++++------
 1 file changed, 23 insertions(+), 6 deletions(-)

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index fcabd85..776ed06 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -2909,23 +2909,40 @@ static void quirk_intel_ntb(struct pci_dev *dev)
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x0e08, quirk_intel_ntb);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x0e0d, quirk_intel_ntb);
 
+#define BOOT_MODE_CHARGER "androidboot.mode=charger"
+
+static bool is_it_charger_mode(void)
+{
+	if (strstr(saved_command_line, BOOT_MODE_CHARGER))
+		return true;
+	else
+		return false;
+}
+
+static int pci_disable_dev_pme_poll(struct pci_dev *pdev, void *data)
+{
+	pdev->pme_poll = false;
+	return 0;
+}
 
 /*PCIe port 0 on Cherryview should support runtime PM*/
 static void quirk_pcie_enable_rtpm(struct pci_dev *dev)
 {
+	bool charger_mode = false;
+
 	dev_info(&dev->dev, "enable runtime PM\n");
+	/* if boot mode is charger OS ignore the children for PCIE0 */
+	charger_mode = is_it_charger_mode();
+	if (charger_mode) {
+		pm_suspend_ignore_children(&dev->dev, true);
+		pci_walk_bus(dev->subordinate, pci_disable_dev_pme_poll, NULL);
+	}
 	pm_runtime_put_noidle(&dev->dev);
 	pm_runtime_allow(&dev->dev);
 }
 DECLARE_PCI_FIXUP_ENABLE(PCI_VENDOR_ID_INTEL,
 	PCI_DEVICE_ID_INTEL_CHV_PCIe_0, quirk_pcie_enable_rtpm);
 
-static int pci_disable_dev_pme_poll(struct pci_dev *pdev, void *data)
-{
-	pdev->pme_poll = false;
-	return 0;
-}
-
 /*PCIEe port 1 on Cherryview should support runtime PM and ignore children*/
 static void quirk_pcie_enable_rtpm_ignore_children(struct pci_dev *dev)
 {
-- 
1.9.1

